algo/src/main/java/org/neo4j/graphalgo/impl/HugeLabelPropagation.java

```diff
@@ -0,0 +1,240 @@
+                nodeProperties.hugeNodeProperties(PARTITION_TYPE),
+                nodeProperties.hugeNodeProperties(WEIGHT_TYPE),
+                batchSize,
+                concurrency,
+                executor,
+                tracker);
+    }
+
+    @Override
+    HugeLabelArray initialLabels(final long nodeCount, final AllocationTracker tracker) {
+        return new HugeLabelArray(HugeLongArray.newArray(nodeCount, tracker));
+    }
+
+    @Override
+    List<BaseStep> baseSteps(
+            final HugeGraph graph,
+            final HugeLabelArray labels,
+            final HugeWeightMapping nodeProperties,
+            final HugeWeightMapping nodeWeights,
+            final Direction direction,
+            final boolean randomizeOrder) {
+        return ParallelUtil.readParallel(
+                concurrency,
+                batchSize,
+                graph,
+                executor,
+                (nodeOffset, nodeIds) -> {
+                    InitStep initStep2 = new InitStep(
+                            nodeProperties,
+                            labels.labels,
+                            nodeIds,
+                            graph,
+                            nodeWeights,
+                            getProgressLogger(),
+                            direction,
+                            randomizeOrder
+                    );
+                    return asStep(initStep2);
+                });
+    }
+
+    private static final class InitStep extends Initialization {
+
+        private final HugeWeightMapping nodeProperties;
+        private final HugeLongArray existingLabels;
+        private final PrimitiveLongIterable nodes;
+        private final HugeGraph graph;
+        private final HugeWeightMapping nodeWeights;
+        private final ProgressLogger progressLogger;
+        private final Direction direction;
+        private final boolean randomizeOrder;
+
+        private InitStep(
+                HugeWeightMapping nodeProperties,
+                HugeLongArray existingLabels,
+                PrimitiveLongIterable nodes,
+                HugeGraph graph,
+                HugeWeightMapping nodeWeights,
+                ProgressLogger progressLogger,
+                Direction direction,
+                boolean randomizeOrder) {
+            this.nodeProperties = nodeProperties;
+            this.existingLabels = existingLabels;
+            this.nodes = nodes;
+            this.graph = graph;
+            this.nodeWeights = nodeWeights;
+            this.progressLogger = progressLogger;
+            this.direction = direction;
+            this.randomizeOrder = randomizeOrder;
+        }
+
+        @Override
+        void setExistingLabels() {
+            PrimitiveLongIterator iterator = nodes.iterator();
+            while (iterator.hasNext()) {
+                long nodeId = iterator.next();
+                long existingLabel = (long) nodeProperties.nodeWeight(nodeId, (double) nodeId);
```

Comment:
could we just keep -1 or such for those nodes and then use the node-id automatically when on read its -1 ?

Reply:
The default for an empty labels array is 0, though we could set all values to -1 with a `setAll` operation more efficiently. Both 0 and -1 could also be actual user-provided values, we don't have any restriction on the partition value. If we restrict the partition to positive values, we could -1 all entries and skip this step.